Add ISO12464 color assessment conditions pop-up window in darkroom

da-phil opened the pull request:
Add ISO12464 color assessment conditions pop-up window in darkroom bottom panel, see screenshot:

![image](https://github.com/user-attachments/assets/fe1b8d1e-aaec-46fc-b810-2b7c84de49e4)


Currently the following config options are hidden in the darktablerc file 
* darkroom/ui/iso12464_border
* darkroom/ui/iso12464_ratio

Therefore it's hard to dial them in, in order to tune the iso12464 assessment borders.
This change adds a pop-up window, which opens up upon right-clicking on the "iso12464 color assessment conditions" button similarly to the overexposure warning pop-up window configuration dialog.

This address the following requests to make those config variables easier to change:
* https://discuss.pixls.us/t/color-assessment-white-border-lots-of-unused-space/42894
* https://github.com/darktable-org/darktable/issues/17278

While tinkering around, I found out that the "iso12464_border" config does not have any influence on the white border, which I find strange.
At least with this change it is easier to debug this issue ;)
I think I'll follow up on this issue once this is merged.


Fixes: https://github.com/darktable-org/darktable/issues/17278

da-phil commented on src/develop/develop.h:
```diff
@@ -329,7 +329,10 @@ typedef struct dt_develop_t
   struct
   {
-    GtkWidget *button; // yes, ugliness is the norm. what did you expect ?
+    GtkWidget *floating_window, *button; // yes, ugliness is the norm. what did you expect ?
```
I really like the comment here :D

da-phil replied:
Will do and replace with a todo comment referring to this refactoring issue: https://github.com/darktable-org/darktable/issues/18559

elstoc commented:
AFAIK there are good reasons for the widths and colours being what they are in this mode and it's supposed to be a technical thing that improves your ability to adjust the contrast/brightness consistently. Putting the setting in darktablerc is us saying "we strongly believe you should leave these settings alone". Honestly I'd prefer the setting wasn't available for adjustment at all, but I can see there might be some screen-specific reasons that it doesn't display as expected. So, if you really need it and you know enough, you can change it.

If we make the adjustment of it so easy that it is presented up-front in the UI this is really us saying to the user "these values aren't important - this is just a UI look-and-feel thing and you should set it to what you find aesthetically pleasing."

While we do allow some border widths to be set in the darkroom in general (and we could consider allowing these to be adjusted more), IMO the colour assessment mode is special and should not be easy to change.

jenshannoschwalm commented:
I fully agree with @elstoc and @TurboGit here. I understand that people on small notebooks think the borders are too wide but that width is simply required as our vision system works that way so it it became a standard defining a minimum border absolute width. Museums for example request that for good reasons as they also request a minimum monitor size.

elstoc commented:
I don't know how much of these parameters are part of the standard, but if we're going to reference a specific ISO number and define an assessment mode as adhering to this standard, then it should either do so, or we should not include the ISO reference number.

da-phil commented:
By the way, does anybody have access to that very ISO standard? Would be very interesting to find out whether it gives very explicit guidance as to how wide the white and neutral grey borders need to be for "optimal" color assessment conditions or even the parameterization dt uses (total border width + white fraction).

> AFAIK there are good reasons for the widths and colours being what they are in this mode and it's supposed to be a technical thing that improves your ability to adjust the contrast/brightness consistently. Putting the setting in darktablerc is us saying "we strongly believe you should leave these settings alone". Honestly I'd prefer the setting wasn't available for adjustment at all, but I can see there might be some screen-specific reasons that it doesn't display as expected. So, if you really need it and you know enough, you can change it.
> 
> If we make the adjustment of it so easy that it is presented up-front in the UI this is really us saying to the user "these values aren't important - this is just a UI look-and-feel thing and you should set it to what you find aesthetically pleasing."
> 
> While we do allow some border widths to be set in the darkroom in general (and we could consider allowing these to be adjusted more), IMO the colour assessment mode is special and should not be easy to change.

If you look at [my post at pixls.us](https://discuss.pixls.us/t/color-assessment-white-border-lots-of-unused-space/42894/15) you will notice that the white & grey frame width changes with the amount of scaling on a high DPI monitor. Overall it gives inconsistent results and sticking to some arbitrarily appearing preset, which I don't believe is specified exactly in that very ISO standard feels a little bit surprising.  
The ideal solution - of course - would be to make this functionality show reproducible results for all resolutions and high DPI scaling factors. But I don't have any idea how to do this, especially the latter.
Hence this functionality, of being able to quickly adapt the color assessment parameters, which enables a more consistent display across my laptop and PC for me.
If someone is happy with the settings, why would they be bothered being able to change them with this pop-up?

> I don't know how many of these parameters are part of the standard, but if we're going to reference a specific ISO number and define an assessment mode as adhering to this standard, then it should either do so, or we should not include the ISO reference number.

If that is a compromise both parties could live with, I'd be more than happy to drop the ISO reference ;)

elstoc commented:
> If that is a compromise both parties could live with, I'd be more than happy to drop the ISO reference ;)

As you might have guessed, that was not really a serious suggestion. By all means let's identify whatever challenges there are in making this work consistently across screen sizes but making it completely free-form basically does away with the whole point of the mode being there in the first place.

jenshannoschwalm commented:
I could not find open access to that iso standard when I implemented the fixed minimum border sizes which was later changed to be somewhat user-adjustable on user request.

But I found recommendations from history imaging museums citating it and explaining how our visual system works. It was one inch each grey/white. Can't find that right now.

da-phil commented:
Apparently also AP didn't have access to the ISO standard when he implemented the feature, see his feature-request back then: https://github.com/darktable-org/darktable/issues/3686 (PR with the feature is linked as well)

So apparently everything is just "eyeballed", not hard numbers from the mentioned standard.

> But I found recommendations from history imaging museums citating it and explaining how our visual system works. It was one inch each grey/white.

At least this would contradict the current default setting of 4 cm total border width (grey + white) and the 0.4 fraction for the white part. How would you calculate those absolute metric numbers for a screen anyway, if there is no way to determine its pixel-pitch (or pixel-per-inch) property. IMHO relative quantities should work much better.

Would be interested in your source from history imaging museums.

jenshannoschwalm commented:
> At least this would contradict the current default setting of 4 cm total border width

Right, that was changed as too many people complained.

> Would be interested in your source from history imaging museums.

Me too. Tried to find is last night but without success. :-(

Weissnix4711 commented:
I have access to everything on BSOL. Ping me an email.

Weissnix4711 commented:
I have done a bit of digging..

For starters, ISO12464 does not exist. It is a typo from 12646.

The standard ISO12646, pertains just to the characteristics (and testing thereof) of monitors. That being things like colour accuracy, viewing angles, surface reflectivity, etc. I'm not sure why this lends its namesake to the darktable feature.

Much more interesting, however, is ISO3664, titled: 'Graphic technologies and photography - Viewing conditions'.

The following recommendations are given about image borders:

> In general, for comparison to prints, which may well be reproduced
> with a white border consisting of unprinted substrate, the border of
> the image should simulate the colour of the substrate;
> for comparison to transparencies, it should be dark.
> However, it is generally preferable that any such border be no more
> than 1 cm to 2 cm wide, even if it would normally be wider on the hardcopy reproduction.

That is all from what I can tell.

Weissnix4711 commented:
> I will recommend sticking with ISO 12646.

At the end of the day, the name of this feature really doesn't matter anyways. Point is, I cannot find any recommendation within any standard that backs up darktable's default values, so I fail to see why we shouldn't allow the user to adjust said values, right? We can change the allowed range to prevent completely ridiculous adjustments, and/or we could add a warning note to whichever dialog the settings may be in, if deemed necessary.

If you know of anything, please let me know. Again, I have access to all of BSOL, incl. all the ISO standards related to graphics, so if the information is there I can verify it. Currently however, I cannot find anything.

jenshannoschwalm commented:
Yes, a big thanks also from my side to find the correct reference.

jenshannoschwalm commented:
Perfect to me.

da-phil commented:
> So let's try to do progress on this.
> 
> Given the recent discussion I'm removing the controversial tag of the PR. And my proposal would be:
> 
>     * Remove the ISO 12646 reference (this seems a source of confusion)
> 
>     * Let's present this feature as **Color Assessment Conditions** only
> 
>     * Remove the `border ratio` from your popup
> 
>     * So keep only the `border width` and allow range  [ 2 .. 5 ], not below 2 (it is currently possible to set it to 0 which makes really no sense to me).
> 
> 
> Before jumping in the implementation I'd like at least the vote of other people working on dt and having commented here (@jenshannoschwalm @elstoc)?

Thanks for given this feature a shot :pray: 

I'm still not sure about the absolute metric value "border width" which currently is impossible to get correct, as it is not possible to get it correct pixel-pitch of the screen (DT currently defaults to 96dpi). I'd be more happy with a relative measure, like a ratio, but I'm open for discussions about a better metric than an absolute value.

Besides that, I need to clean-up the PR a little bit, as I stopped the work after the push-back in the discussions.
I don't want file I/O (update the darktablerc file) in each callback call when the slider(s) are moved. It should only update the current dev context while scrolling the sliders and only update the darktablerc file after the popup window is closed (I don't know which gtk event is used for that yet).

elstoc commented:
IMO if we're going to expose modification to this it would be really useful to have some guidelines on physical/relative widths for the user manual when people adjust it.

da-phil commented:
> Besides that, I need to clean-up the PR a little bit, as I stopped the work after the push-back in the discussions. I don't want file I/O (update the darktablerc file) in each callback call when the slider(s) are moved. It should only update the current dev context while scrolling the sliders and only update the darktablerc file after the popup window is closed (I don't know which gtk event is used for that yet).

Having said that, I just realized that `dt_conf_set_*` functions do not write directly to disk, but just update a key-value store (hash table), and that hash table will be written to the darktablerc file upon shutdown of DT. Therefore I updated my PR (2nd commit)  to just use interaction with the conf-store, instead of introducing new values to the `iso_12646` struct within `develop.h`.
At least updating conf values is guarded by a mutex, which should make the interaction between value-writing in callback and GUI rendering safer. Is that the preferred way to handle gui parameters which happen to be config values too @TurboGit @jenshannoschwalm?


> IMO if we're going to expose modification to this it would be really useful to have some guidelines on physical/relative widths for the user manual when people adjust it.

Sounds like a good idea :+1: 
I'll need to tinker around a little bit more to figure out:
* how to get consistent borders across different monitors (with different DPI resolutions) within darktable, taking @TurboGit  idea of only one parameter into account
* a rough guideline how to set the parameter(s) to get good results for the purpose of assessing colors & brightness of an image
*

da-phil commented on data/darktableconfig.xml.in:
```diff
@@ -2237,16 +2237,16 @@
   </dtconfig>
   <dtconfig>
-    <name>darkroom/ui/iso12464_border</name>
+    <name>darkroom/ui/color_assessment_border</name>
```
@TurboGit how do we make sure existing darktablerc configs are updated with the new config param names?

da-phil replied:
Understood, makes sense :+1:

parafin commented:
> * how to get consistent borders across different monitors (with different DPI resolutions) within darktable, taking @TurboGit  idea of only one parameter into account

DPI settings are available in `struct dt_gui_gtk_t` and there’s also DT_PIXEL_APPLY_DPI helper macros.

da-phil commented:
> > * how to get consistent borders across different monitors (with different DPI resolutions) within darktable, taking @TurboGit  idea of only one parameter into account
> 
> DPI settings are available in `struct dt_gui_gtk_t` and there’s also `DT_PIXEL_APPLY_DPI()` helper macros.

I'm aware, but the values `dpi` seem to be something else than I expected. I'm talking about the physical DPI value, which in case of my laptop screen is 243dpi, similar to Apple retina screens.
The reported `dpi` value in `dt_get_screen_resolution()` is 129dpi, because I'm using a font scaling of 1.35x to make fonts more readable, while keeping the overall display scale at 1x. If I keep 1x as font scaling factor, the function gives me the well known 96dpi.

It's not surprising that the used function `gdk-screen-get-resolution()` seems to be mostly used for font handling, [see documentation](https://www.manpagez.com/html/gdk3/gdk3-3.24.14/GdkScreen.php#gdk-screen-get-resolution):
> Gets the resolution for font handling on the screen; see [gdk_screen_set_resolution()](https://www.manpagez.com/html/gdk3/gdk3-3.24.14/GdkScreen.php#gdk-screen-set-resolution) for full details.

I implemented a new function (`dt_get_physical_screen_dpi`) which gives me the physical dpi value and this one reliably gives me 243dpi and correct metric border widths of 1-5cm (verified with a ruler on the screen).

It would be great if people with different screens could try out the current state of the PR to see if they always get consistent results across different screens ;)

parafin commented:
Modern DEs make this too complicated. Your “font scaling” may be just overriding the reported DPI. I’m not sure if it’s “correct” behavior to ignore that override. You should try maybe first to handle the simple case where all scaling is set to 1. You should get correct DPI and correct physical size on screen. If DPI is incorrect, then your DE is doing bad things. Or you run macOS, where they just decided that DPI is always 96 (Apple really like to go their own way, I guess out of spite).

da-phil commented:
> Modern DEs make this too complicated. Your “font scaling” may be just overriding the reported DPI. I’m not sure if it’s “correct” behavior to ignore that override. You should try maybe first to handle the simple case where all scalins are set to 1. You should get correct DPI and correct physical size on screen. If DPI is incorrect, then your DE is doing bad things. Or you run macOS, where they just decided that DPI is always 96 (Apple really like to go their own way, I guess out of spite).

I think it all makes sense, I just wasn't aware of all this before working on the changes in this PR.
As I quoted the gtk documentation above about `gdk-screen-get-resolution()`, the purpose of it is mainly for font scaling (not a physical quantity).
I was looking for something else: the physical resolution of the screen (pixels per centimeters), which you can use to get physically correct measurements on a screen, based on a length in centimeters.
However, I'd rather like to avoid the current approach of setting the total border width between 2-5cm, as this might not make sense for really big desktop screens, a relative quantity, which scales between small laptop screens to large desktop monitors, might make more sense overall.

da-phil commented:
@TurboGit I'm proposing to stick to the current approach of having two parameters, one for the total border width and one for the white border fraction. However now I made the total border width relative to the screen size, which should scale even better between small (laptop) and big desktop or TV screens.

This is how the dialog looks now and the default settings should be very close to the previous ones. 
![Screenshot 2025-03-22 01-37-18](https://github.com/user-attachments/assets/b3fe73b9-b2be-443c-8a71-6f45c55fc69f)

Please have a look and provide some feedback.

Remaining ToDos:
* add "right click for options" tooltip text (similar to other popup windows) to remaining translations
* small clean-up of stuff I came across

da-phil commented:
In the meantime I also finished the translations of the additional text in the tooltip msg.

I'm still waiting for some feedback. Is everybody with how the borders are rendered at different screen sizes and resolutions?

victoryforce commented:
> In the meantime I also finished the translations of the additional text in the tooltip msg.

Please remove the translations. You shouldn't do it yourself, that's the translators' job.

da-phil commented:
> > In the meantime I also finished the translations of the additional text in the tooltip msg.
> 
> Please remove the translations. You shouldn't do it yourself, that's the translators' job.

Because this translation change was trivial for all languages, as this sentence was already present in all available language files, I thought it would be convenient to not wait for all the individual translations to be updated by the translators.

da-phil commented:
> Well we still want the translator to double check the strings so I agree better remove all the po files.

Should I also remove the po changes where I removed the ISO12464 reference?

da-phil commented:
> Works for me, a very minor clean-up proposed. Some nice clean-up. Thanks!

Cool, thanks for the feedback.
Anybody else eager to try it out and see if it works for you? @elstoc @jenshannoschwalm @parafin @victoryforce

jenshannoschwalm commented:
Fine with me.

da-phil commented:
@TurboGit release notes: 
* Color assessment conditions: Changed total border width to relative scaling which should work well on small and big screens, independent of physical screen resolution, added pop-up window for parameterization and removed ISO12464 reference
